docs/rfcs/030-vectored-timeline-get.md

```diff
@@ -0,0 +1,78 @@
+# Vectored Timeline Get
+
+Created on: 2024-01-02
+Author: Christian Schwarz
+
+# Summary
+
+A brief RFC / GitHub Epic describing a vectored version of the `Timeline::get` method that is at the heart of Pageserver.
+
+# Motivation
+
+During basebackup, we issue many `Timeline::get` calls for SLRU pages that are *adjacent* in key space.
+For an example, see
+https://github.com/neondatabase/neon/blob/5c88213eaf1b1e29c610a078d0b380f69ed49a7e/pageserver/src/basebackup.rs#L281-L302.
+
+Each call must traverse the layer map to gather reconstruct data (`Timeline::get_reconstruct_data`) for the requested page number (`blknum` in the example).
+
+That is quite inefficient because:
+
+1. We do the layer map traversal repeatedly, even if, e.g., all the data sits in the same image layer at the bottom of the stack.
+2. Anecdotally, keys adjacent in keyspace and written simultaneously also end up physically adjacent in the layer files [^1].
+   So, to provide the reconstruct data for N adjacent keys, we would actually only _need_ to issue a single large read to the filesystem, instead of the N reads we currently do.
+   The filesystem, in turn, ideally stores the layer file physically contiguously, so our large read will turn into one IOP toward the disk.
+
+[^1]: https://www.notion.so/neondatabase/Christian-Investigation-Slow-Basebackups-Early-2023-12-34ea5c7dcdc1485d9ac3731da4d2a6fc?pvs=4#15ee4e143392461fa64590679c8f54c9
+
+# Solution
+
+We should have a vectored aka batched aka scatter-gather style alternative API for `Timeline::get`. Having such an API  unlocks:
+
+* more efficient basebackup
+* batched IO during compaction (useful for strides of unchanged pages)
+* page_service: expose vectored get_page_at_lsn for compute (=> good for seqscan / prefetch)
+  * if [on-demand SLRU downloads](https://github.com/neondatabase/neon/pull/6151) land before vectored Timeline::get, on-demand SLRU downloads will still benefit from this API
+
+# DoD
+
+There is a new variant of `Timeline::get`, called `Timeline::get_vectored`.
+It takes as arguments an `lsn: Lsn` and a `src: &[KeyVec]` where `struct KeyVec { base: Key, count: usize }`.
+
+It is up to the implementor to figure out a suitable and efficient way to return the reconstructed page images.
+It is sufficient to simply return a `Vec<Bytes>`, but, likely more efficient solutions can be found after studying all the callers of `Timeline::get`.
```

Comment:
It seems like the high level question here is whether the result should be a stream or not.  i.e. whether it's up to the caller to ensure they issue "reasonably" small vectored gets, or whether the caller can issue a vectored get for an entire relation and then receive a stream.

Without having thought about it deeply, I would lean toward a non-streaming API, and some simple code in places like the basebackup handler to constrain the number of keys we fetch at once.

Reply:
Yeah, thought about stream of `Bytes` as well.

Other idea would be to make the caller pre-allocate the buffer and pass in a `&mut [u8]`.

But, if we want to reduce memcopys, that's a bad design because io_uring will need an owned buffer.
Also ties in with the question about future of the page cache.

Do you think it makes sense to decide on `Vec<Bytes>` vs `Stream<Bytes>` now or see how things pan out during implementation?

Reply:
Now that we have a bit more detail in the implementation section below, I think we definitely will want to bound the number of items in a vectored read, in order to walk the index once without using excessive memory to carry all the key state.  Let's keep the streaming part very high up the stack: an unlimited-size streaming vectored get will just decompose into a series of vectored gets, and we can punt on whether we need to implement the streaming wrapper.